### DISCUSSION TITLE
Choosing between CoreCLR and Mono

### DISCUSSION THREAD

**Discussion by stefan-sf-ibm** [dotnet/runtime#59650] [Q&A]
Is there any way to choose between CoreCLR and Mono? I tried adding `UseMonoRuntime` to my csproj file without luck, i.e., CoreCLR is still used.

I'm asking because ultimately I would like to run a couple of BenchmarkDotnet tests while switching between CoreCLR and Mono.

I'm using dotnet 6.0 RC1 on x64.

**Comment by danmoseley** [comment]
@steveisok

**Comment by steveisok** [comment]
@stefan-sf-ibm Can you share a branch / files?

**Comment by stefan-sf-ibm** [comment]
I started of a hello world build by doing `dotnet new console` and added `UseMonoRuntime` to the csproj file:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <UseMonoRuntime>true</UseMonoRuntime>
  </PropertyGroup>

</Project>
```

Running `MONO_ENV_OPTIONS="-v -v -v -v" dotnet run` doesn't output anything. Running the same while using the systems mono binary prints a lot of debug information. Thus I assumed that the prior command still uses CoreCLR instead of Mono. Did I miss something?

**Comment by steveisok** [comment]
Ah, looks like this will run using the shared framework, which will not contain any of the mono parts.  It's something that is technically possible, but since all mono configurations are self contained (w/ dotnet publish), we haven't had the need to go all the way. 

Can you try `dotnet publish` and run your executable from there? You should get the mono runtime at that point.

**Comment by stefan-sf-ibm** [comment]
Running `dotnet publish` and afterwards `MONO_ENV_OPTIONS="-v -v -v -v" ./bin/Debug/net6.0/publish/tt` still doesn't print any debug output.

**Comment by steveisok** [comment]
I should have asked earlier.  What OS?

**Comment by stefan-sf-ibm** [comment]
I'm on Fedora 33

**Comment by PathogenDavid** [comment]
> I'm asking because ultimately I would like to run a couple of BenchmarkDotnet tests while switching between CoreCLR and Mono.

BenchmarkDotNet automatically rebuilds your benchmark and launches that instead of using the one you built. (I've never looked into why it does this, but I've always assumed it has to do with how it instruments things for benchmarking.) The situation might be that you're starting the BenchmarkDotNet runner as Mono but it's starting the actual benchmarks with CoreCLR.

If your goal is to compare benchmarks between CoreCLR and Mono, you should probably use [BenchmarkDotNet's built-in support for doing so](https://benchmarkdotnet.org/articles/samples/IntroJobBaseline.html).

**Comment by stefan-sf-ibm** [comment]
@PathogenDavid thanks for the hint! I will give it a try :)
